src/blockrelay/mempool_sync.cpp

```diff
@@ -0,0 +1,444 @@
+    }
+
+    LOG(MPOOLSYNC, "Received getmemsynctx from peer=%s requesting %d transactions\n", pfrom->GetLogName(),
+        reqMempoolSyncTx.setCheapHashesToRequest.size());
+
+    std::vector<uint256> mempoolTxHashes;
+    GetMempoolTxHashes(mempoolTxHashes);
+
+    // Locate transactions requested
+    // Note that only those still in the mempool will be located
+    std::vector<CTransactionRef> vTx;
+    for (auto &hash : mempoolTxHashes)
+    {
+        uint64_t cheapHash = GetShortID(
+            mempoolSyncResponded[pfrom].shorttxidk0, mempoolSyncResponded[pfrom].shorttxidk1, hash, SHORT_ID_VERSION);
+
+        if (reqMempoolSyncTx.setCheapHashesToRequest.count(cheapHash) == 0)
+            continue;
+
+        auto txRef = mempool.get(hash);
+        if (txRef == nullptr)
+            continue;
+
+        vTx.push_back(txRef);
+    }
+
+    LOG(MPOOLSYNC, "Sending %d mempool sync transactions to peer=%s\n", vTx.size(), pfrom->GetLogName());
+
+    // Assemble missing transaction object
+    CMempoolSyncTx mempoolSyncTx(vTx);
+    pfrom->PushMessage(NetMsgType::MEMPOOLSYNCTX, mempoolSyncTx);
+
+    // We should not receive any future messages related to this synchronization round
+    mempoolSyncResponded[pfrom].completed = true;
+
+    return true;
+}
+
+bool CMempoolSyncTx::HandleMessage(CDataStream &vRecv, CNode *pfrom)
+{
+    std::string strCommand = NetMsgType::MEMPOOLSYNCTX;
+    CMempoolSyncTx mempoolSyncTx;
+    vRecv >> mempoolSyncTx;
+
+    // Do not process unrequested memsynctx.
+    if (mempoolSyncRequested.count(pfrom) == 0)
+    {
+        dosMan.Misbehaving(pfrom, 10);
+        return error("Received memsynctx from peer %s but mempool sync is not in progress", pfrom->GetLogName());
+    }
+
+    // Already received requested transactions
+    if (mempoolSyncRequested[pfrom].completed)
+    {
+        dosMan.Misbehaving(pfrom, 100);
+        return error("Received memsynctx from peer %s but transactions have already been sent", pfrom->GetLogName());
+    }
+
+    LOG(MPOOLSYNC, "Received memsynctx from peer=%s; adding %d transactions to mempool\n", pfrom->GetLogName(),
+        mempoolSyncTx.vTx.size());
+
+    // Add transactions to mempool
+    for (const auto &tx : mempoolSyncTx.vTx)
+    {
+        CTxInputData inputData;
+        inputData.tx = tx;
+        inputData.nodeId = pfrom->id;
+        EnqueueTxForAdmission(inputData);
+    }
+
+    LOG(MPOOLSYNC, "Recovered %d txs from peer=%s via mempool sync\n", mempoolSyncTx.vTx.size(), pfrom->GetLogName());
+
+    // We should not receive any future messages related to this round of synchronization
+    mempoolSyncRequested[pfrom].completed = true;
+
+    return true;
+}
+
+void GetMempoolTxHashes(std::vector<uint256> &mempoolTxHashes)
+{
+    {
+        READLOCK(orphanpool.cs_orphanpool);
+        for (auto &kv : orphanpool.mapOrphanTransactions)
+        {
+            mempoolTxHashes.push_back(kv.first);
+        }
+    }
```

Comment:
don't you have to add the commitQ hashes as well here? 

Also, why not use a set instead of a vector in case you get duplicates which could happen if for instance you remove the orphan pool lock and then the orphan slips into the mempool.